### ISSUE TITLE
v2.6.1 breaks existing test - `Assert.Equal() Failure: Collections differ`

### ISSUE THREAD

**Issue by tig** [xunit/xunit#2821]
I have not been able to create a simpler repro yet but will if it's needed.

We have the following test:

```cs
Assert.Equal (TextModel.StringToRuneCells ("Threally really first line."), tm.GetLine (0));
```

With v2.5.3 this test passed.

With xUnit v2.6.1 we get:

```
Xunit.Sdk.EqualException: 'Assert.Equal() Failure: Collections differ
           ↓ (pos 0)
Expected: [U+0054 'T'; null, U+0068 'h'; null, U+0072 'r'; null, U+0065 'e'; null, U+0061 'a'; null, ···]
Actual:   [U+0054 'T'; null, U+0068 'h'; null, U+0072 'r'; null, U+0065 'e'; null, U+0061 'a'; null, ···]
           ↑ (pos 0)'
```

Breaking on this Assert and viewing the two objects in Immediate Window, they are the same:

```
tm.GetLine(0)
Count = 27
    [0]: "U+0054 'T'; null"
    [1]: "U+0068 'h'; null"
    [2]: "U+0072 'r'; null"
    [3]: "U+0065 'e'; null"
    [4]: "U+0061 'a'; null"
    [5]: "U+006C 'l'; null"
    [6]: "U+006C 'l'; null"
    [7]: "U+0079 'y'; null"
    [8]: "U+0020 ' '; null"
    [9]: "U+0072 'r'; null"
    [10]: "U+0065 'e'; null"
    [11]: "U+0061 'a'; null"
    [12]: "U+006C 'l'; null"
    [13]: "U+006C 'l'; null"
    [14]: "U+0079 'y'; null"
    [15]: "U+0020 ' '; null"
    [16]: "U+0066 'f'; null"
    [17]: "U+0069 'i'; null"
    [18]: "U+0072 'r'; null"
    [19]: "U+0073 's'; null"
    [20]: "U+0074 't'; null"
    [21]: "U+0020 ' '; null"
    [22]: "U+006C 'l'; null"
    [23]: "U+0069 'i'; null"
    [24]: "U+006E 'n'; null"
    [25]: "U+0065 'e'; null"
    [26]: "U+002E '.'; null"
TextModel.StringToRuneCells ("Threally really first line.")
Count = 27
    [0]: "U+0054 'T'; null"
    [1]: "U+0068 'h'; null"
    [2]: "U+0072 'r'; null"
    [3]: "U+0065 'e'; null"
    [4]: "U+0061 'a'; null"
    [5]: "U+006C 'l'; null"
    [6]: "U+006C 'l'; null"
    [7]: "U+0079 'y'; null"
    [8]: "U+0020 ' '; null"
    [9]: "U+0072 'r'; null"
    [10]: "U+0065 'e'; null"
    [11]: "U+0061 'a'; null"
    [12]: "U+006C 'l'; null"
    [13]: "U+006C 'l'; null"
    [14]: "U+0079 'y'; null"
    [15]: "U+0020 ' '; null"
    [16]: "U+0066 'f'; null"
    [17]: "U+0069 'i'; null"
    [18]: "U+0072 'r'; null"
    [19]: "U+0073 's'; null"
    [20]: "U+0074 't'; null"
    [21]: "U+0020 ' '; null"
    [22]: "U+006C 'l'; null"
    [23]: "U+0069 'i'; null"
    [24]: "U+006E 'n'; null"
    [25]: "U+0065 'e'; null"
    [26]: "U+002E '.'; null"
```

Are there any known issues in xUnit > v2.5.3 that might be causing this?

**Comment by bradwilson** [comment]
@Tig I will need a full repro to track this down.

**Comment by tig** [comment]
> @tig I will need a full repro to track this down.

With xUnit 2.5.3 `TestObj.Equals` gets called.

With xUnit 2.6.1 it does not.
```cs
		public class TestObj : IEquatable<TestObj> {
			public char Char { get; set; }

			public bool Equals (TestObj other)
			{
				return other?.Char == Char;
			}
		}


		[Fact]
		public void Repo_xUnit_2821 ()
		{
			List<TestObj> list1 = new List<TestObj> ();
			list1.Add (new TestObj { Char = 'a' });

			List<TestObj> list2 = new List<TestObj> ();
			list2.Add (new TestObj { Char = 'a' });

			Assert.Equal(list1, list2);
		}
```

**Comment by bradwilson** [comment]
This looks like a pretty strange edge case that came about because of [performance optimizations](https://github.com/xunit/assert.xunit/commit/6a94a86a02cd23c9235a412dde156f8ad73503cb). We were sort of accidentally getting the right result for the wrong reason previously. Working on a fix now.

**Comment by bradwilson** [comment]
Available in v2: `2.6.2-pre.15`
Available in v3: `0.1.1-pre.325`

**Comment by tig** [comment]
> 2.6.2-pre

Will you push this to nuget?

![image](https://github.com/xunit/xunit/assets/585482/783a8f68-ac18-40f2-aebe-ffbe81554c5f)

**Comment by bradwilson** [comment]
Not until a final RTM. You can use CI builds from the feedz.io feed, instructions here: https://xunit.net/docs/using-ci-builds

**Comment by bradwilson** [comment]
@gioce90 Yes, the bug specified here was fixed in 2.6.2.

The new bug you opened is a duplicate of a bug that is already fixed and will be shipped in 2.6.3.
